ls1intum/Artemis
# Feature/avoid duplicating programming submissions

### Checklist

### Motivation and Context


Programming exercise submissions were duplicated  and we wanted to avoid this my assigning a commit to a single submission. Jenkins and Bamboo services contained duplicated code which can be extracted in an abstract class.

### Description

We made the following changes:

- Commits are now assigned to submissions. The `ProgrammingSubmissionService` checks if a submissions is already assigned to a certain commit and returns either the existing one or creates a new submission.
- The OTHER submission type is removed and replaced by MANUAL because it's not needed and was a workaround for creating fallback submissions. 
- `JenkinsService` and `BambooService` common methods have been extracted to a new `AbstractContinuousIntegrationService` super class which contains common behavior between both services
- We updated integration tests to reflect the changes.
- Rename the "Result" header to "Last Result" on the scores and participation pages
- Add a "Result Count" in the submission page. It shows how much results a submission has

### Steps for Testing

Please test on both Bamboo and Jenkins test servers!
1. Participate in a programming exercise and make sure the build results are working
2. While you're in the code editor make one or two submissions.
2. Go into course management -> the programming exercise -> participations
3. Make sure the number of submissions is the amount of times you pressed the submit button
3. Re-trigger a participation by clicking the circle arrow button. The submission number shouldn't change
4. Press the "Trigger All" button. The submission number shouldn't change
5. Press the "Trigger Failed" button. The submission number shouldn't change
6. Change the exercise due date so that it's in the past (done)
6. Navigate to course management -> assessment and assess the submission
7. Add some feedback, submit it and make sure the feedback is displayed to the user
8. Change some test cases in the test repository, change some weights in the grading page
8. After doing some manual assessments, run again the "Trigger All" button and make sure the manual assessments are still valid. Also make sure the automatic feedback elements in the manual assessments are updated according to the test case and weights changes 

### Test Coverage




### Screenshots
![Screenshot from 2021-01-19 17-33-53](https://user-images.githubusercontent.com/11298674/105064181-dae7c900-5a7c-11eb-9053-9f6f5ab2cab9.png)
![Screenshot from 2021-01-19 17-34-33](https://user-images.githubusercontent.com/11298674/105064185-db805f80-5a7c-11eb-8965-af9c0db2f420.png)
